lib/internal/http2/core.js

```diff
@@ -214,18 +222,14 @@ function onSessionTrailers(id) {
          'Internal HTTP/2 Failure. Stream does not exist. Please ' +
          'report this as a bug in Node.js');
 
-  // TODO(jasnell): mapToHeaders will throw synchronously if the headers
-  // are not valid. Try catch to keep it from bubbling up to the native
-  // layer so that we can emit. mapToHeaders can be refactored to take
-  // an optional callback or event emitter instance so it can emit errors
-  // async instead.
-  try {
-    const trailers = Object.create(null);
-    stream.emit('fetchTrailers', trailers);
-    return mapToHeaders(trailers, assertValidPseudoHeaderTrailer);
-  } catch (err) {
-    process.nextTick(() => stream.emit('error', err));
+  const trailers = Object.create(null);
+  stream.emit('fetchTrailers', trailers);
+  const headersList = mapToHeaders(trailers, assertValidPseudoHeaderTrailer);
+  if (headersList instanceof Error) {
```

Comment:
mapToHeaders might return an Array or an Error (yes, I know polymorphic returns are awful). I'm thinking that maybe we can do a less expensive check here.